### ISSUE TITLE
use case: 1:1 pool replication

### ISSUE THREAD

**Issue by problame** [zrepl/zrepl#330]
Hi,

The case for zfs recv -d is very simple, we want to replicate a full pool from one HA node to the other (which is not exactly HA, because we have a 1-2 minute delay between both nodes due to replication, but allow us to have both nodes with separate pools able to serve NFS requests and failover very quickly without having to re-import a pool).
The current zrepl logic always tries to receive a pool into a predefined destination (which forbids us the use of "sink" and a push logic), and when pulling datasets, it always tries to recv with the source pool name into our destination.
Example:
source node with pool sourcepool/*
dest node with pool destpool
pull job always creates destpool/sourcepool/*  and there is no way to receive directly sourcepool/* into destpool/*

I agree that instead of using the -d flag we could change the logic implemented into the "sink" or "pull" jobs to have the same behaviour, but it's currently not possible to replicate a full pool without having the pool name into the destination.

_Originally posted by @gbarazer in https://github.com/zrepl/zrepl/pull/285#issuecomment-637382904_

**Comment by problame** [comment]
I assume your root filesystem is _not_ on the `sourcepool` or `destpool` (i.e. most likely non-zfs rootfs?)

**Comment by gbarazer** [comment]
Of course :)

**Comment by problame** [comment]
Is the replica pool read-only?

It would be relatively easy (few hours of work, incl. testing) to add config options that allow for a behavior similar to `-d` in zfs recv.

But for the time being, zrepl assumes that it 'owns' the receiving side filesystems, i.e. nothing is writing to them.
If something does write to the receiving side, creates snapshots, etc, then zrepl would stop replicating because it detects a replication conflict.

**Comment by InsanePrawn** [comment]
Related? #221

**Comment by danderson** [comment]
Just to contribute a motivating use case: I'm in the middle of upgrading my storage server, and I'm doing a pool replacement: I have `data` and `new-data` pools, and I need to replicate the data + snapshots over. Once I'm done, I will do a cutover by hand and retire the old pool.

Using zrepl for this would be great, because it's just the best user experience for handling movement of data between ZFS targets. However, because I can't set the target names explicitly, I would end up with snapshots under `new-data/$source_name/data/*`, which I would then need to clone and promote back to their correct names. This means I lose the history of snapshots on the new datasets (and thus my pruning policies no longer apply), and once the migration is complete I end up with "remnants" of the migration in the form of standalone snapshots for the old dataset.

Some kind of 1:1 replication configuration would address this, since I could make zrepl replicate `data<` to `new-data<`, and end up with an exact duplicate of the original dataset structure in the new pool.

With all that said, I realize that this is not a perfect fit for zrepl's use case of being a replication tool for backups. I would be effectively using it as a much better/robust commandline wrapper around zfs-send/zfs-recv, to make use of its ability to plan and execute a complex replication strategy of many snapshots, with resumability, etc. etc.

Of course, I can do the same thing with raw zfs send/receive, but it's much more annoying to do, since I have 1014 snapshots to replicate. Additionally, while I'm doing this replication, zrepl is still running on the original pool and is taking new snapshots, and applying grid pruning to old ones. zrepl replication handles this completely seamlessly and handles snapshots appearing/vanishing gracefully, and it would be very nice to leverage that.
